File: server/src/intTest/java/org/tctalent/server/repository/db/UserRepositoryIntTest.java
/*
 * Copyright (c) 2024 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.repository.db;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedUser;

import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.tctalent.server.model.db.Role;
import org.tctalent.server.model.db.Status;
import org.tctalent.server.model.db.User;
import org.tctalent.server.repository.db.integrationhelp.BaseDBIntegrationTest;

public class UserRepositoryIntTest extends BaseDBIntegrationTest {

  @Autowired
  private UserRepository repo;
  private User user;

  @BeforeEach
  public void setup() {
    assertTrue(isContainerInitialised());
    user = getSavedUser(repo);
  }

  // This also tests for not deleted!!
  @Test
  public void testFindByUsernameAndRole() {
    User u = repo.findByUsernameAndRole(user.getUsername(), user.getRole());
    assertNotNull(u);
    assertEquals(user.getId(), u.getId());

    // add in deleted just in case the query changes.
    user.setStatus(Status.deleted);
    repo.save(user);
    u = repo.findByUsernameAndRole(user.getUsername(), user.getRole());
    assertNull(u);
  }

  @Test
  public void testFindByUsernameIgnoreCase() {
    User u = repo.findByUsernameIgnoreCase(user.getUsername());
    assertNotNull(u);
    assertEquals(user.getId(), u.getId());

    // add in deleted just in case the query changes.
    user.setStatus(Status.deleted);
    repo.save(user);
    u = repo.findByUsernameIgnoreCase(user.getUsername());
    assertNull(u);
  }

  @Test
  public void testFindByEmailIgnoreCase() {
    User u = repo.findByEmailIgnoreCase(user.getEmail());
    assertNotNull(u);
    assertEquals(user.getId(), u.getId());

    user.setStatus(Status.deleted);
    repo.save(user);
    u = repo.findByEmailIgnoreCase(user.getEmail());
    assertNull(u);
  }

  @Test
  public void testFindByResetToken() {
    user.setResetToken("YES");
    repo.save(user);
    User u = repo.findByResetToken(user.getResetToken());
    assertNotNull(u);
    assertEquals(user.getId(), u.getId());

    user.setStatus(Status.deleted);
    repo.save(user);
    u = repo.findByResetToken(user.getResetToken());
    assertNull(u);
  }

  @Test
  public void testSearchAdminUsersName() {
    user.setRole(Role.admin);
    repo.save(user);
    Page<User> u = repo.searchAdminUsersName(user.getUsername(), Pageable.unpaged());
    assertNotNull(u);

    // Will be content of 1 user in the page.
    assertEquals(1, u.getContent().size());
  }

  // This is an unusual query - data modified to make the test work.
  @Test
  public void testSearchAdminUsersNameFail() {
    Page<User> u = repo.searchAdminUsersName(user.getUsername(), Pageable.unpaged());
    assertNotNull(u);
    // Will be no users in the content.
    assertEquals(0, u.getContent().size());
  }

  @Test
  public void testSearchStaffNotUsingMfa() {
    user.setRole(Role.admin);
    repo.save(user);
    List<User> u = repo.searchStaffNotUsingMfa();
    assertNotNull(u);
    assertEquals(1, u.size());
    List<Long> ids = u.stream().map(User::getId).toList();
    assertEquals(user.getId(), ids.getFirst());
  }

  @Test
  public void testSearchStaffNotUsingMfaFailUsing() {
    user.setUsingMfa(true);
    repo.save(user);
    List<User> u = repo.searchStaffNotUsingMfa();
    assertNotNull(u);
    assertTrue(u.isEmpty());
  }

  @Test
  public void testSearchStaffNotUsingMfaFailDeleted() {
    user.setStatus(Status.deleted);
    repo.save(user);
    List<User> u = repo.searchStaffNotUsingMfa();
    assertNotNull(u);
    assertTrue(u.isEmpty());
  }
}

File: server/src/intTest/kotlin/org/tctalent/server/repository/db/UserRepositoryIntTest.kt
/*
 * Copyright (c) 2024 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.repository.db

import org.junit.jupiter.api.Assertions.assertNotNull
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Pageable
import org.tctalent.server.model.db.Role
import org.tctalent.server.model.db.Status
import org.tctalent.server.model.db.User
import org.tctalent.server.repository.db.integrationhelp.BaseDBIntegrationTest
import org.tctalent.server.repository.db.integrationhelp.getSavedUser
import kotlin.test.*

class UserRepositoryIntTest : BaseDBIntegrationTest() {
  @Autowired lateinit var repo: UserRepository
  private lateinit var user: User

  @BeforeTest
  fun setup() {
    assertTrue { isContainerInitialized() }
    user = getSavedUser(repo)
  }

  // This also tests for not deleted!!
  @Test
  fun `test find by username and role`() {
    var u = repo.findByUsernameAndRole(user.username, user.role)
    assertNotNull(u)
    assertEquals(user.id, u.id)

    // add in deleted just in case the query changes.
    repo.save(user.apply { status = Status.deleted })
    u = repo.findByUsernameAndRole(user.username, user.role)
    assertNull(u)
  }

  @Test
  fun `test find by username ignorecase`() {
    var u = repo.findByUsernameIgnoreCase(user.username)
    assertNotNull(u)
    assertEquals(user.id, u.id)

    // add in deleted just in case the query changes.
    repo.save(user.apply { status = Status.deleted })
    u = repo.findByUsernameIgnoreCase(user.username)
    assertNull(u)
  }

  @Test
  fun `test find by email ignorecase`() {
    var u = repo.findByEmailIgnoreCase(user.email)
    assertNotNull(u)
    assertEquals(user.id, u.id)

    repo.save(user.apply { status = Status.deleted })
    u = repo.findByEmailIgnoreCase(user.email)
    assertNull(u)
  }

  @Test
  fun `test find by reset token`() {
    repo.save(user.apply { resetToken = "YES" })
    var u = repo.findByResetToken(user.resetToken)
    assertNotNull(u)
    assertEquals(user.id, u.id)

    repo.save(user.apply { status = Status.deleted })
    u = repo.findByResetToken(user.resetToken)
    assertNull(u)
  }

  @Test
  fun `test search admin users name`() {
    repo.save(user.apply { role = Role.admin })
    val u = repo.searchAdminUsersName(user.username, Pageable.unpaged())
    assertNotNull(u)

    // Will be content of 1 user in the page.
    assertEquals(1, u.content.size)
  }

  // This is an unusual query - data modified to make the test work.
  @Test
  fun `test search admin users name fail`() {
    val u = repo.searchAdminUsersName(user.username, Pageable.unpaged())
    assertNotNull(u)
    // Will be no users in the content.
    assertEquals(0, u.content.size)
  }

  @Test
  fun `test search staff not using mfa`() {
    repo.save(user.apply { role = Role.admin })
    val u = repo.searchStaffNotUsingMfa()
    assertNotNull(u)
    assertEquals(1, u.size)
    val ids = u.map { it.id }
    assertEquals(user.id, ids.first())
  }

  @Test
  fun `test search staff not using mfa fail using`() {
    repo.save(user.apply { usingMfa = true })
    val u = repo.searchStaffNotUsingMfa()
    assertNotNull(u)
    assertTrue { u.isEmpty() }
  }

  @Test
  fun `test search staff not using mfa fail deleted`() {
    repo.save(user.apply { status = Status.deleted })
    val u = repo.searchStaffNotUsingMfa()
    assertNotNull(u)
    assertTrue { u.isEmpty() }
  }
}
